// @ngneat/lib/src/utils/files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setJSON = exports.getJSONFromFS = exports.getJSON = exports.writeFile = exports.readFile = exports.readFileFromFS = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const fs = require("fs");
function throwPathNotFound(path) {
    throw new schematics_1.SchematicsException(`Could not find ${path}`);
}
function readFileFromFS(path) {
    return fs.readFileSync(path, { encoding: "utf-8" });
}
exports.readFileFromFS = readFileFromFS;
function readFile(host, path) {
    const buffer = host.read(path);
    if (buffer) {
        return buffer.toString("utf-8");
    }
    throwPathNotFound(path);
}
exports.readFile = readFile;
function writeFile(path, content, host) {
    host.overwrite(path, content);
}
exports.writeFile = writeFile;
function getJSON(host, path) {
    if (!host.exists(path)) {
        throwPathNotFound(path);
    }
    const sourceText = readFile(host, path);
    return JSON.parse(sourceText);
}
exports.getJSON = getJSON;
function getJSONFromFS(path) {
    if (!fs.existsSync(path)) {
        throwPathNotFound(path);
    }
    const sourceText = readFileFromFS(path);
    return JSON.parse(sourceText);
}
exports.getJSONFromFS = getJSONFromFS;
function setJSON(host, path, content) {
    host.overwrite(path, JSON.stringify(content, null, 2));
    return host;
}
exports.setJSON = setJSON;
