// @ngneat/lib/src/lib/add-spectator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addSpectator = void 0;
const path = require("path");
const schematics_1 = require("@angular-devkit/schematics");
function addSpectator(options, scopeWithName) {
    return (tree, context) => {
        context.logger.info("⌛ Adding @ngneat/spectator...");
        const module = tree.actions.find((action) => !!action.path.match(/\.module\.ts/));
        if (!module) {
            return schematics_1.noop();
        }
        const component = schematics_1.externalSchematic("@ngneat/spectator", "spectator-component", {
            path: path.dirname(module.path),
            name: options.name,
            skipImport: true,
            flat: true,
            inlineStyle: true,
            inlineTemplate: true,
            project: scopeWithName,
        });
        const service = schematics_1.externalSchematic("@ngneat/spectator", "spectator-service", {
            path: path.dirname(module.path),
            name: options.name,
        });
        return schematics_1.chain([component, service]);
    };
}
exports.addSpectator = addSpectator;
